### DISCUSSION TITLE
RPi4 and Coral USB: standard vs maximum speed

### DISCUSSION THREAD

**Discussion by notgood** [blakeblackshear/frigate#3512] [Ideas]
I'm running Frigate on Raspberry Pi 4B and obviosly running into USB power issues, so a bit of simple math:

```
RPi4 maximum total USB peripheral current draw:         1.2 A
SSD max current draw:                                   0.7 A
Google Coral USB Edge TPU reduced current draw:         0.5 A
Google Coral USB Edge TPU max current draw:             0.9 A
```
Quick grep inside Frigate container says its using libedgetpu1-max, pushing Coral USB into max current draw 900ma
So sooner or later SSD write and Coral detection happens at the same time and RPi4 USB overcurrent polyfuses resets ports, killing either SSD or Coral, or both.

Now my question is, perhaps is could be feasible to make Frigate use libedgetpu1-max or libedgetpu1-std, depending on user preference? In my case it would allow me to avoid headache and trouble of finding compatible externally powered hub.

**Comment by NickM-27** [comment]
I'm not sure what the inference times are for the standard runtime and there may be other trade offs as well. 

The docs highly recommend a powered USB hub as the coral is very temperamental when it comes to power either way

**Comment by notgood** [comment]
Yeah, Coral is very power temperamental, probably spiking behind specs.
And RPi4 USB power is badly engineered, providing four USB ports and only 1.2A total is very questionable. 
RPi4 itself consumes 1.3A under max stress, so USB should be given remaining 1.7A, but alas...

Powered USB3 hub is another can of worms:
1) Aesthetics of another box and bunch of wires, yey.
2) Good luck finding decently built USB3 hub, that doesn't backfeed external power to RPi4 and correctly switches between power sources, instead of sharing load with both external and internal 5V.
3) Try guessing which chipset does your random USB3 hub comes with, some are good and some aren't.
4) Another PSU to worry about, more connections to fail.

Another solution would be to power Coral from RPi4 +5V pin, as it is connected directly to 3A PSU without any polyfuses. But that would require custom USB3 cable with a split power wire. Not sure such thing is available on market, and DIY soldering and injecting would be messy because of USB3 wire count.

**Reply by NickM-27** [reply]
That all sounds more likely to be having issues than just using a hub. We have one linked in the docs that some users have tried and reported success https://docs.frigate.video/installation/#raspberry-pi-34

**Reply by notgood** [reply]
> Another solution would be to power Coral from RPi4 +5V pin, as it is connected directly to 3A PSU without any polyfuses.

This turned out to be quite easy:
1) Remove Coral USB plastic case
2) Solder standard jumper wire to +5V testpoints, these are quite big and easy to solder.
3) Connect other side of jumper wire to RPi4 +5V pin.
4) Disable USB cable +5V using piece of tape to avoid backfeeding USB ports.

Frigate is working stable for weeks, RPi4 stock PSU is capable enough to feed both SSD and Coral in max power mode.

Oh, and let me shame RPi4 engineers again for artificial limit on USB power. 
I also disagree with Frigate recomendation on USB3 hub usage, its reduces reliability of the whole setup.

![155734](https://user-images.githubusercontent.com/8156459/184641714-a9cd3c1f-bfa2-4c0c-8881-00528ed21d48.jpg)
![153718](https://user-images.githubusercontent.com/8156459/184641734-3a382fe1-ec95-4143-82d5-f5d7d4fc4ddb.jpg)

**Reply by purplefis** [reply]
> > Another solution would be to power Coral from RPi4 +5V pin, as it is connected directly to 3A PSU without any polyfuses.
> 
> This turned out to be quite easy:
> 
> 1. Remove Coral USB plastic case
> 2. Solder standard jumper wire to +5V testpoints, these are quite big and easy to solder.
> 3. Connect other side of jumper wire to RPi4 +5V pin.
> 4. Disable USB cable +5V using piece of tape to avoid backfeeding USB ports.
> 
> Frigate is working stable for weeks, RPi4 stock PSU is capable enough to feed both SSD and Coral in max power mode.
> 
> Oh, and let me shame RPi4 engineers again for artificial limit on USB power. I also disagree with Frigate recomendation on USB3 hub usage, its reduces reliability of the whole setup.
> 
> ![155734](https://user-images.githubusercontent.com/8156459/184641714-a9cd3c1f-bfa2-4c0c-8881-00528ed21d48.jpg) ![153718](https://user-images.githubusercontent.com/8156459/184641734-3a382fe1-ec95-4143-82d5-f5d7d4fc4ddb.jpg)

Can u confirm this is still working stable?

**Reply by norbertjoni** [reply]
> > Another solution would be to power Coral from RPi4 +5V pin, as it is connected directly to 3A PSU without any polyfuses.
> 
> 
> 
> This turned out to be quite easy:
> 
> 1) Remove Coral USB plastic case
> 
> 2) Solder standard jumper wire to +5V testpoints, these are quite big and easy to solder.
> 
> 3) Connect other side of jumper wire to RPi4 +5V pin.
> 
> 4) Disable USB cable +5V using piece of tape to avoid backfeeding USB ports.
> 
> 
> 
> Frigate is working stable for weeks, RPi4 stock PSU is capable enough to feed both SSD and Coral in max power mode.
> 
> 
> 
> Oh, and let me shame RPi4 engineers again for artificial limit on USB power. 
> 
> I also disagree with Frigate recomendation on USB3 hub usage, its reduces reliability of the whole setup.
> 
> 
> 
> ![155734](https://user-images.githubusercontent.com/8156459/184641714-a9cd3c1f-bfa2-4c0c-8881-00528ed21d48.jpg)
> 
> ![153718](https://user-images.githubusercontent.com/8156459/184641734-3a382fe1-ec95-4143-82d5-f5d7d4fc4ddb.jpg)
> 
> 
> 
> 

Hi friend,

It's still works stable? How many cameras can you run on it? With HA or is this just for a frigate? 

Thank you!

**Reply by LaurenceGough** [reply]
My Pi 4 is stable with Coral and USB Sata 3, no hub. I've noticed SSD power draw can differ greatly between brands. I am using the Integral V series 120GB. A decent, quality power supply is essential.

**Reply by norbertjoni** [reply]
> My Pi 4 is stable with Coral and USB Sata 3, no hub. I've noticed SSD power draw can differ greatly between brands. I am using the Integral V series 120GB. A decent, quality power supply is essential. 

How many cameras do you run? HA and Frigate how much CPU use?

**Reply by LaurenceGough** [reply]
Just two but coral can be constantly in use if it's foggy or snowing / heavy rain.  Ha and frigate yes. Probably about 70% peak AVG around 30-40 percent.  But updates of containers etc peak it higher or at max.

**Reply by bob454522** [reply]
I too would like to see a built in option for the user to choose libedgetpu1-std (max can remain the default) - inside of frigate, wo having to manual modify the container, which will break on frigate updates).   In my case im running 2x containers, each with its on usb coral (+ a p1000 gpu helping with h264).   
tks

**Comment by FrigateEnthusiast** [comment]
Google docs warn about using the max driver and heat. I would just connect the coral thumb inside the case in a good air flow section and not worry about it.

**Comment by notgood** [comment]
> Hi friend,
> 
> It's still works stable? How many cameras can you run on it? With HA or is this just for a frigate?
> 
> Thank you!

1.5 year later answer is yes, with a direct wired feed of +5v from RPi4 to Coral whole setup is stable. One camera only, just Frigate container.

System load average: 0.96, 0.68, 0.71
